File: pkg/src/cwt_morlet.c

/***************************************************************
*              (c) Copyright  1997                             *
*                         by                                   *
*     Author: Rene Carmona, Bruno Torresani, Wen-Liang Hwang   *
*                 Princeton University                         *
*                 All right reserved                           *
***************************************************************/

#include "Swave.h"
#include "complex.h"


/***************************************************************
*  Function: multi:
*  ---------
*     Multiplication of 2 vectors in the Fourier domain; the
*     first one is complex-valued, as well as the output; the
*     second one is real-valued.
*
*   Ri1, Ii1: first input vector.
*   Ri2: second input vector (real).
*   Or, Oi: output vector.
*   isize: length of the vectors.
***************************************************************/

void multi(double *Ri1, double *Ii1, double *Ri2, double *Or,
  double *Oi, int isize)
{
  int i;

  for(i = 0; i < isize; i++) {
    Or[i] = Ri1[i] * Ri2[i];
    Oi[i] = Ii1[i] * Ri2[i];
  }
  return;
}


/***************************************************************
*  Function: morlet_frequency:
*  ---------
*     Generates a Morlet wavelet in the frequency domain.
*     The wavelet is centered at the origin, and normalized so
*     that psi(0) =1/sqrt(2 pi)
*
*   w: wavelet
*   scale: scale of the wavelet
*   isize: window size
*   cf: central frequency of the wavelet
***************************************************************/

void morlet_frequency(float cf,float scale,double *w,int isize)
{
  double tmp, tmp1=0;
  int i;
  double twopi;

  twopi = 6.28318530717959;
  
/*  tmp1 = exp(-(cf  * cf)/2); */
  for(i = 0; i < isize; i++) {
    tmp = (double)(scale * i * twopi/isize - cf);
    tmp = -(tmp * tmp)/2;
/*    w[i] = exp(tmp) - tmp1; */
    w[i] = exp(tmp);
  }
  return;
}

/***************************************************************
*  Function: morlet_time:
*  ---------
*     Generates a Morlet wavelet in the time domain.
*     The wavelet is centered at the origin, and normalized so
*     that psi(0) = 1
*
*   w: wavelet
*   scale: scale of the wavelet
*   isize: window size
*   cf: central frequency of the wavelet
*
* remark: unlike the other functions, this one generates an 
*         array starting at 1, for compatibility with S. 
***************************************************************/

void morlet_time(float *pcf,float *pscale, int *pb, 
		 double *w_r, double *w_i,int *pisize)
{
  double tmp, tmp2;
  float cf = *pcf, scale = *pscale;
  int b = *pb, isize = *pisize;
  int i;

  for(i = 1; i <= isize; i++) {
    tmp = (double)((double)(i-b)/scale); 
    tmp2 = exp(-(tmp * tmp)/2.);
    w_r[i-1] = tmp2*cos(tmp*cf)/scale;
    w_i[i-1] = tmp2*sin(tmp*cf)/scale;
  }
  return;
}



/***************************************************************
*  Function: vmorlet_time:
*  ---------
*     Generates Morlet wavelets located on nodes of the ridge,
*     in the time domain.
*     The mother wavelet is centered at the origin, and normalized so
*     that psi(0) = 1/sqrt(2 pi)
*
*   w_r, w_i: wavelet (real and imaginary parts)
*   scale: scale of the wavelets on the ridge samples
*   b: position of the wavelets on the ridge samples
*   isize: window size
*   cf: central frequency of the wavelet
*   nbnodes: number of ridge samples
*
* remark: unlike the other functions, this one generates an 
*         array starting at 1, for compatibility with S. 
***************************************************************/
void vmorlet_time(float *pcf,float *pscale, int *b, 
		 double *w_r, double *w_i,int *pisize, int *pnbnode)
{
  double tmp, tmp2;
  double cf = (double)(*pcf);
  int isize = *pisize;
  int i, j, nbnode = *pnbnode;
  int position;
  double sqtwopi;

  sqtwopi = sqrt(6.28318530717959);

  for(j = 0; j < nbnode; j++) {
         position = b[j];
     for(i = 1; i <= isize; i++) {
         tmp = (double)((double)(i-position)/pscale[j]); 
         tmp2 = exp(-(tmp * tmp)/2.)/pscale[j]/sqtwopi;
         w_r[j * isize + i-1] = tmp2*cos(tmp*cf);
         w_i[j * isize + i-1] = tmp2*sin(tmp*cf);
      }
   }
  return;
}


/*****************************************************************
*  function:  Scwt_morlet_r
*    Continuous wavelet transform :
*
*   input: (real-valued) input signal
*   Ri1, Ii1: Fourier transform of input signal (real and
*      imaginary parts).
*   Ri2: Real part of Fourier transform of Morlet wavelet
*   Oreal,Oimage: real and imaginary parts of CWT
*   pinputsize: signal size
*   pnboctave: number of scales (powers of 2)
*   pnvoice: number of scales between 2 consecutive powers of 2
*   pcenterfrequency: centralfrequency of Morlet wavelet
******************************************************************/

void Scwt_morlet_r(float *input, double *Oreal, double *Oimage,
   int *pnboctave, int *pnbvoice, int *pinputsize, float *pcenterfrequency)
{	
  int nboctave, nbvoice, i, j, inputsize;
  float centerfrequency, a;
  double *Ri2, *Ri1, *Ii1, *Ii, *Ri;


  centerfrequency = *pcenterfrequency;
  nboctave = *pnboctave;
  nbvoice = *pnbvoice;
  inputsize = *pinputsize;
  if(!(Ri2 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri2 in cwt_morlet.c \n");
  if(!(Ri1 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri1 in cwt_morlet.c \n");
  if(!(Ii1 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ii1 in cwt_morlet.c \n");
  if(!(Ri = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri in cwt_morlet.c \n");
  if(!(Ii = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ii in cwt_morlet.c \n");

  for(i = 0; i < inputsize; i++) {
    Ri[i] = (double)input[i]; 
    Ii[i] = 0.0;
  }

  double_fft(Ri1,Ii1,Ri,Ii,inputsize,-1);   
  
  for(i = 1; i <= nboctave; i++) {
    for(j=0; j < nbvoice; j++) {
      a = (float)(pow((double)2,(double)(i+j/((double)nbvoice))));
      morlet_frequency(centerfrequency,a,Ri2,inputsize); 
      multi(Ri1,Ii1,Ri2,Oreal,Oimage,inputsize);
      double_fft(Oreal,Oimage,Oreal,Oimage,inputsize,1); 
      Oreal = Oreal + inputsize;
      Oimage = Oimage + inputsize;  
    }
  }

  return;
}




/*****************************************************************
*  function:  Scwt_morlet
*    Continuous wavelet transform :
*
*   input: (a priori complex-valued) input signal
*   Ri1, Ii1: Fourier transform of input signal (real and
*      imaginary parts).
*   Ri2: Real part of Fourier transform of Morlet wavelet
*   Oreal,Oimage: real and imaginary parts of CWT
*   pinputsize: signal size
*   pnboctave: number of scales (powers of 2)
*   pnvoice: number of scales between 2 consecutive powers of 2
*   pcenterfrequency: centralfrequency of Morlet wavelet
******************************************************************/

void Scwt_morlet(float *Rinput,float *Iinput,double *Oreal,
   double *Oimage,int *pnboctave,int *pnbvoice,
   int *pinputsize,float *pcenterfrequency)
{	
  int nboctave, nbvoice, i, j, k, inputsize;
  float centerfrequency, a;
  double *Ri2, *Ri1, *Ii1, *Ii, *Ri;


  centerfrequency = *pcenterfrequency;
  nboctave = *pnboctave;
  nbvoice = *pnbvoice;
  inputsize = *pinputsize;
  if(!(Ri2 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri2 in cwt_morlet.c \n");
  if(!(Ri1 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri1 in cwt_morlet.c \n");
  if(!(Ii1 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ii1 in cwt_morlet.c \n");
  if(!(Ri = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri in cwt_morlet.c \n");
  if(!(Ii = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ii in cwt_morlet.c \n");

  for(i = 0; i < inputsize; i++) {
    Ri[i] = (double)Rinput[i]; 
    Ii[i] = (double)Iinput[i];
  }

  double_fft(Ri1,Ii1,Ri,Ii,inputsize,-1);   
  
  for(i = 1; i <= nboctave; i++) {
    for(j=0; j < nbvoice; j++) {
      a = (float)(pow((double)2,(double)(i+j/((double)nbvoice))));
      morlet_frequency(centerfrequency,a,Ri2,inputsize); 
      multi(Ri1,Ii1,Ri2,Oreal,Oimage,inputsize);
      double_fft(Oreal,Oimage,Oreal,Oimage,inputsize,1); 
      Oreal = Oreal + inputsize;
      Oimage = Oimage + inputsize;  
    }
  }

}



/***************************************************************
*  Function: Svwt_morlet:
*  ---------
*     Computes the continuous wavelet transform of input
*     signal with Morlet wavelet, a fixed scale a
*
*     Rinput, Iinput: real and imaginary parts of the signal
*     Oreal, Oimage: real and imaginary parts of the cwt.
***************************************************************/

void Svwt_morlet(float *Rinput,float *Iinput,double *Oreal,
   double *Oimage,float *pa,int *pinputsize,
   float *pcenterfrequency)
{	
  int octave, voice, nbvoice, i, j, k, inputsize;
  float centerfrequency, a;
  double *Ri2, *Ri1, *Ii1, *Ii, *Ri;


  centerfrequency = *pcenterfrequency;
  a = *pa;
  inputsize = *pinputsize;
  if(!(Ri2 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri2 in cwt_morlet.c \n");
  if(!(Ri1 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri1 in cwt_morlet.c \n");
  if(!(Ii1 = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ii1 in cwt_morlet.c \n");
  if(!(Ri = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ri in cwt_morlet.c \n");
  if(!(Ii = (double *)  R_alloc(inputsize, sizeof(double)) ))
    error("Memory allocation failed for Ii in cwt_morlet.c \n");

  for(i = 0; i < inputsize; i++) {
    Ri[i] = (double)Rinput[i]; 
    Ii[i] = (double)Iinput[i];
  }

  double_fft(Ri1,Ii1,Ri,Ii,inputsize,-1);   
  
  morlet_frequency(centerfrequency,a,Ri2,inputsize); 
  multi(Ri1,Ii1,Ri2,Oreal,Oimage,inputsize);
  double_fft(Oreal,Oimage,Oreal,Oimage,inputsize,1); 
  
}

File: branches/Rcwt/R/Cwt_Morlet.R
#########################################################################
#      $Log: Cwt_Morlet.S,v $
#########################################################################
#
#               (c) Copyright  1997                             
#                          by                                   
#      Author: Rene Carmona, Bruno Torresani, Wen-Liang Hwang   
#                  Princeton University
#                  All right reserved                           
#########################################################################

cwt <- function(input, noctave, nvoice = 1, w0 = 2*pi, twoD = TRUE,
                plot = TRUE)
#########################################################################
#      cwt:
#      ---
#       continuous wavelet transform function
# 	compute the continuous wavelet transform with (complex-valued)
#			Morlet wavelet
#
#       Input:
#       ------
# 	 input: input signal (possibly complex-valued)
#	 noctave: number of powers of 2 for the scale variable
#	 nvoice: number of scales between 2 consecutive powers of 2
#        w0: central frequency of Morlet wavelet
#	 twoD: if set to TRUE, organizes the output as a 2D array 
#			(signal_size X nb_scales)
#		      if not: 3D array (signal_size X noctave X nvoice)
#	 plot: if set to TRUE, displays the modulus of cwt on the graphic
#		device.
#
#       Output:
#       -------
#        tmp: continuous (complex) wavelet transform
#
#########################################################################
{
  oldinput <- input
  isize <- length(oldinput)
  
  tmp <- adjust.length(oldinput)
  input <- tmp$signal
  newsize <- length(input)
  
  pp <- noctave * nvoice
  Routput <- matrix(0,newsize,pp)
  Ioutput <- matrix(0,newsize,pp)
  output <- matrix(0,newsize,pp)
  dim(Routput) <- c(pp * newsize,1)
  dim(Ioutput) <- c(pp * newsize,1)
  dim(input) <- c(newsize,1)
  
  z <- .C("Scwt_morlet",
          as.double(Re(input)),
          as.double(Im(input)),
          Rtmp = as.double(Routput),
          Itmp = as.double(Ioutput),
          as.integer(noctave),
          as.integer(nvoice),
          as.integer(newsize),
          as.double(w0),
          PACKAGE="Rcwt")
  
  Routput <- z$Rtmp
  Ioutput <- z$Itmp
  dim(Routput) <- c(newsize,pp)
  dim(Ioutput) <- c(newsize,pp)
  if(twoD) {
    output <- Routput[1:isize,] + 1i*Ioutput[1:isize,]
    if(plot) {
      image(Mod(output), xlab="Time", ylab="log(scale)",
            main="Wavelet Transform Modulus")
    }
    output
  } 
  else {
    Rtmp <- array(0,c(isize,noctave,nvoice))
    Itmp <- array(0,c(isize,noctave,nvoice))
    for(i in 1:noctave)
      for(j in 1:nvoice) {
        Rtmp[,i,j] <- Routput[1:isize,(i-1)*nvoice+j]
        Itmp[,i,j] <- Ioutput[1:isize,(i-1)*nvoice+j]
      }
    Rtmp + 1i*Itmp
  }
}


morlet <- function(sigsize, location, scale, w0 = 2*pi)
#########################################################################
#       morlet: Morlet's wavelet  
#       -------
#        Generates a Morlet wavelet for given location and scale
#
#       input:
#       ------
#        sigsize: signal size (dimension of the array)
#        location: location of the wavelet
#        scale: value of the scale at which the transform is computed
#        w0: central frequency of Morlet wavelet
#
#       output:
#       -------
#        z$wavelet.r + z$wavelet.i * i: wavelet (complex 1D array
#          of size sigsize)
#
#########################################################################
{
  wavelet.r <- numeric(sigsize)
  wavelet.i <- numeric(sigsize)
  
  z <- .C("morlet_time",
          as.double(w0),
          as.double(scale),
          as.integer(location),
          wavelet.r = as.double(wavelet.r),
          wavelet.i = as.double(wavelet.i),
          as.integer(sigsize),
          PACKAGE="Rcwt")
  
  z$wavelet.r + 1i*z$wavelet.i
}
